akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.routing
+import akka.actor.typed._
+import akka.actor.typed.receptionist.Receptionist
+import akka.actor.typed.receptionist.ServiceKey
+import akka.actor.typed.scaladsl.AbstractBehavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.annotation.InternalApi
+import akka.dispatch.forkjoin.ThreadLocalRandom
+
+/**
+ * Provides builder style configuration options for group routers while still being a behavior that can be spawned
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class GroupRouterBuilder[T] private[akka] (
+  key:          ServiceKey[T],
+  logicFactory: () ⇒ RoutingLogic[T] = () ⇒ RoutingLogics.randomLogic[T]()
+) extends javadsl.GroupRouter[T]
+  with scaladsl.GroupRouter[T] {
+
+  // deferred creation of the actual router
+  def apply(ctx: TypedActorContext[T]): Behavior[T] = new GroupRouterImpl[T](ctx.asScala, key, logicFactory())
+
+  def withRandomRouting(): GroupRouterBuilder[T] = copy(logicFactory = RoutingLogics.randomLogic[T] _)
+
+  def withRoundRobinRouting(): GroupRouterBuilder[T] = copy(logicFactory = () ⇒ new RoutingLogics.RoundRobinLogic[T])
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private final class GroupRouterImpl[T](
+  ctx:          ActorContext[T],
+  serviceKey:   ServiceKey[T],
+  routingLogic: RoutingLogic[T]
+) extends AbstractBehavior[T] {
+
+  private var routees: Array[ActorRef[T]] = Array.empty[ActorRef[T]]
+
+  // casting trix to avoid having to wrap incoming messages - note that this will cause problems if intercepting
+  // messages to a router
+  ctx.system.receptionist ! Receptionist.Subscribe(serviceKey, ctx.self.unsafeUpcast[Any].narrow[Receptionist.Listing])
+
+  def onMessage(msg: T): Behavior[T] = msg match {
+    case serviceKey.Listing(update) ⇒
+      // we don't need to watch, because receptionist already does that
+      val newRoutees = update.toArray
+      if (newRoutees.nonEmpty) {
+        // make sure we keep a somewhat similar order, but don't put all entries from
+        // the same node next to each other
+        newRoutees.sortBy(ref ⇒ (ref.path.toStringWithoutAddress, ref.path.address))
```

Comment:
This wouldn't be needed for the random logic. Should we make the logic stateful also with regards to the `routees`, then the logic itself can decide such things as sorting and it would also solve the API wart of using Array (in case we make the `RoutingLogic` public later. The logic can convert to Array internally when routees are updated. The signatures could look something like this:

```
trait RoutingLogic[T] {
  def routeesUpdated(newRoutees: immutable.Seq[ActorRef[T]]): Unit
  def hasRoutees: Boolean
  def selectRoutee(): ActorRef[T]
}
```

Reply:
Hmm, but we'd still need to keep the set of routees or we couldn't create the `newRoutees` for pool when a child terminates.

Reply:
It's a bit tricky to make a logic API that works nicely with both removing a single actor (pool) and updating the whole set (group). But maybe it could just have one method for each of those.

Reply:
For pool we know live routees via context.children and could pass in that as newRoutees when one Terminates

Reply:
Ah, clever, didn't think of that. I'll try with this API then, I don't think we need `hasRoutees` though because the behavior will already know if the children or last update from receptionist is empty.